### PR TITLE
[.NET] Turkish DateTime DateTimePeriod support

### PR DESCRIPTION
**Pull request by LionbridgeCSII**
All test cases pass.

### CODE REVIEW

File: .NET/Microsoft.Recognizers.Text.DateTime/Turkish/Parsers/TurkishDateTimePeriodParserConfiguration.cs [hunk 1]
```diff
@@ -24,7 +24,8 @@ public class TurkishDateTimePeriodParserConfiguration : BaseDateTimeOptionsConfi
         public TurkishDateTimePeriodParserConfiguration(ICommonDateTimeParserConfiguration config)
             : base(config)
         {
-            TokenBeforeDate = DateTimeDefinitions.TokenBeforeDate;
+            // @TODO replace token with list of tokens (in Turkish many different suffixes can be used)
+            TokenBeforeDate = DateTimeDefinitions.TokenBeforeDate1;
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Turkish/Parsers/TurkishDateTimePeriodParserConfiguration.cs]: Will this happen before this PR can merge?

File: .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateTimePeriodExtractor.cs [hunk 2]
```diff
@@ -133,8 +148,9 @@ private IEnumerable<Token> MergeDateWithTimePeriodSuffix(string text, List<Extra
         // Cases like "today after 2:00pm", "1/1/2015 before 2:00 in the afternoon"
         // Valid connector in English for Before include: "before", "no later than", "in advance of", "prior to", "earlier than", "sooner than", "by", "till", "until"...
         // Valid connector in English for After include: "after", "later than"
-        private bool IsValidConnectorForDateAndTimePeriod(string text)
+        private bool IsValidConnectorForDateAndTimePeriod(string text, out int length, bool inPrefix)
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateTimePeriodExtractor.cs]: If you need the length as output, I'd suggest returning the length, and return an error code (-1) when false.
You can declare a Constants entry for this. Maybe INVALID_CONNECTOR_CODE.

Comment (tellarin) [reply .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateTimePeriodExtractor.cs]: It's OK to rename such method to GetValidConnectorIndexForDateAndTimePeriod
and declare a new IsValidConnectorForDateAndTimePeriod that just checks the returned index or code and still returns a bool.

File: .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateTimePeriodExtractor.cs [hunk 3]
```diff
@@ -637,5 +672,41 @@ private List<Token> MatchRelativeUnit(string text)
...
+                    var beforeStr = text.Substring(0, duration.Start);
+                    var matchNext = this.config.NextPrefixRegex.Match(beforeStr);
+                    success = match.Success || matchNext.Success;
+                    if (MatchPrefixRegexInSegment(beforeStr, matchNext, true))
+                    {
+                        startToken = matchNext.Index;
+                    }
+                }
+
+                if (success)
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateTimePeriodExtractor.cs]: What happens if not a success?

Comment (LionbridgeCSII) [reply .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateTimePeriodExtractor.cs]: if not a success, the method returns a token that starts and ends at -1. When the method is called, it is checked if the returned token.Start is >=0.

File: .NET/Microsoft.Recognizers.Text.DateTime/Turkish/Parsers/TurkishDateTimePeriodParserConfiguration.cs (hunk 1 above)
Comment (LionbridgeCSII) [reply .NET/Microsoft.Recognizers.Text.DateTime/Turkish/Parsers/TurkishDateTimePeriodParserConfiguration.cs]: yes, updated this and modified BaseDateTimePeriodParser to work with a list of tokens (when before/after flag is true)

File: .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateTimePeriodExtractor.cs (hunk 2 above)
Comment (LionbridgeCSII) [reply .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateTimePeriodExtractor.cs]: (the length is needed only when the before/after flag is true).
Modified the method to return the length. 
When the method is called, it is checked if the returned length is >=0 (as for the method in the comment below)

File: .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateTimePeriodExtractor.cs [hunk 4]
```diff
@@ -148,9 +150,9 @@ private IEnumerable<Token> MergeDateWithTimePeriodSuffix(string text, List<Extra
         // Cases like "today after 2:00pm", "1/1/2015 before 2:00 in the afternoon"
         // Valid connector in English for Before include: "before", "no later than", "in advance of", "prior to", "earlier than", "sooner than", "by", "till", "until"...
         // Valid connector in English for After include: "after", "later than"
-        private bool IsValidConnectorForDateAndTimePeriod(string text, out int length, bool inPrefix)
+        private int GetValidConnectorForDateAndTimePeriod(string text, bool inPrefix)
         {
-            length = 0;
+            int length = -1;
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateTimePeriodExtractor.cs]: Please declare a constant for it, and use it here and in the return checks.

File: Patterns/Turkish/Turkish-DateTime.yaml [hunk 5]
```diff
@@ -123,7 +123,7 @@ DateUnitRegex: !simpleRegex
 DateTokenPrefix: ''
 TimeTokenPrefix: ''
 TokenBeforeDate: ''
```
Comment (tellarin) [inline Patterns/Turkish/Turkish-DateTime.yaml]: Can you add a comment here in the YAML file explaining why TokenBeforeDate needs to stay empty and TokenListBeforeDate populated for DateTimePeriod?

File: .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateTimePeriodExtractor.cs (hunk 4 above)
Comment (LionbridgeCSII) [reply .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateTimePeriodExtractor.cs]: done.

File: Patterns/Turkish/Turkish-DateTime.yaml (hunk 5 above)
Comment (LionbridgeCSII) [reply Patterns/Turkish/Turkish-DateTime.yaml]: Comment added.

**Pull request merged**